### PR TITLE
Initial security hardening doc

### PR DESCRIPTION
**Pull request by ale8k**

### CODE REVIEW

File: how-to/security_hardening.rst [hunk 1]
```diff
@@ -0,0 +1,60 @@
+JAAS: Security Hardening
+=======================
+JIMM, the service at the centre of JAAS can be hardened in a number of ways. This 
+document details how you can harden the security of your JAAS deployment. 
+
+CORS
+----
+To set CORS on JIMM, use the configuration option "cors-allowed-origins".
```
Comment (alesstimec) [inline how-to/security_hardening.rst]: perhaps a link to where the reader can read about CORS

File: how-to/security_hardening.rst [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+LEGO & NGINX Ingress Integrator
```
Comment (alesstimec) [inline how-to/security_hardening.rst]: suggestion: change the title.. perhaps just `TLS`

File: how-to/security_hardening.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+With JAAS deployed, you can deploy both LEGO and the Inregator, and integrate your LEGO charm deployment
+to your ingress integrator, and then the ingress integrator to JIMM to enable TLS ingress for
+
+Identity Provider
+-----------------
+JAAS uses the Canonical Identity Platform for authentication. The communication between JAAS
+and the identity platform can be secured via TLS.
+
+You will require the identity Platform and the self-signed-certificates charm deployed.
```
Comment (alesstimec) [inline how-to/security_hardening.rst]: hmm ideally they should have proper cert, perhaps using the httprequest-lego-k8s charm.. you may `suggest` using the self-signed-certificates charm for local deploy, but definitely not for production

File: how-to/security_hardening.rst [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+See `here <https://charmhub.io/topics/canonical-identity-platform/tutorials/e2e-tutorial>`__ for deploying the identity platform. 
+
+Your identity platform will require TLS enabled via the `self signed certificates charm <https://charmhub.io/self-signed-certificates>`__.
+
+Using JIMM's receive-ca-cer integration, you can now relate to the self-signed-certificates charm
```
Comment (alesstimec) [inline how-to/security_hardening.rst]: s/receive-ca-cer/receive-ca-cert/

File: how-to/security_hardening.rst [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+With JAAS deployed, you can deploy both LEGO and the Inregator, and integrate your LEGO charm deployment
+to your ingress integrator, and then the ingress integrator to JIMM to enable TLS ingress for
+
+Identity Provider
+-----------------
+JAAS uses the Canonical Identity Platform for authentication. The communication between JAAS
+and the identity platform can be secured via TLS.
+
+You will require the identity Platform and the self-signed-certificates charm deployed.
+See `here <https://charmhub.io/topics/canonical-identity-platform/tutorials/e2e-tutorial>`__ for deploying the identity platform.
```
Comment (alesstimec) [inline how-to/security_hardening.rst:31]: this guide seems to suggest the self signed certs.. maybe that's enough and we just add the bits that are needed to relate jimm to self-signed-certs

File: .custom_wordlist.txt [hunk 6]
```diff
@@ -12,6 +12,7 @@ CLI
 CN
 config
 CORS
+cors
```
Comment (babakks) [inline .custom_wordlist.txt]: This shouldn't be here.

File: how-to/security_hardening.rst (hunk 1 above)
Comment (ale8k) [reply how-to/security_hardening.rst]: i left a little para and link

File: .custom_wordlist.txt [hunk 7]
```diff
@@ -31,6 +32,8 @@ HMAC
 https
 IAM
 IdP
+integrator
+Integrator
```
Comment (babakks) [inline .custom_wordlist.txt]: And these two.

File: how-to/security_hardening.rst (hunk 2 above)
Comment (ale8k) [reply how-to/security_hardening.rst]: But this is ingress tls only?I changed it to ingress tls

File: how-to/index.rst [hunk 8]
```diff
@@ -31,3 +31,11 @@ Observability
...
 
    Integrate with the Canonical Observability Stack <integrate_with_cos>
+
+Security
+--------
+
+.. toctree::
+   :maxdepth: 1
+
+   Harden your JIMM deployment <security_hardening>
```
Comment (babakks) [inline how-to/index.rst]: Let's remove "your" from here.

File: how-to/security_hardening.rst (hunk 3 above)
Comment (ale8k) [reply how-to/security_hardening.rst]: this is in-cluster tohugh isnt it??

File: how-to/security_hardening.rst (hunk 5 above)
Comment (ale8k) [reply how-to/security_hardening.rst:31]: responded on other comment

File: how-to/security_hardening.rst [hunk 9]
```diff
@@ -0,0 +1,60 @@
+JAAS: Security Hardening
+========================
+JIMM, the service at the centre of JAAS can be hardened in a number of ways. This 
+document details how you can harden the security of your JAAS deployment. 
+
+CORS
+----
+To set CORS on JIMM, use the configuration option "cors-allowed-origins".
```
Comment (babakks) [inline how-to/security_hardening.rst]: Surround `cors-allowed-origins` with two backticks like this and you won't need to update the `.custom_wordlist.txt` file:

```
``cors-allowed-origins``
```

File: how-to/security_hardening.rst [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+With JAAS deployed, you can deploy both LEGO and the integrator, and integrate your LEGO charm deployment
+to your ingress integrator, and then the ingress integrator to JIMM to enable TLS ingress for
+
+Identity Provider
+-----------------
+JAAS uses the Canonical Identity Platform for authentication. The communication between JAAS
+and the identity platform can be secured via TLS.
+
+You will require the identity Platform and the self-signed-certificates charm deployed.
```
Comment (babakks) [inline how-to/security_hardening.rst]: 1. Surround `self-signed-certificates` with double backticks.
2. s/identity/Identity

File: how-to/security_hardening.rst [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+See `here <https://charmhub.io/topics/canonical-identity-platform/tutorials/e2e-tutorial>`__ for deploying the identity platform. 
+
+Your identity platform will require TLS enabled via the `self signed certificates charm <https://charmhub.io/self-signed-certificates>`__.
```
Comment (babakks) [inline how-to/security_hardening.rst]: s/self signed/self-signed

File: how-to/security_hardening.rst [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+Using JIMM's receive-ca-cert integration, you can now relate to the self-signed-certificates charm
```
Comment (babakks) [inline how-to/security_hardening.rst]: Surround `receive-ca-cert` and `self-signed-certificates` with double backticks.

File: how-to/security_hardening.rst [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+to enabled TLS between the identity platform and JIMM.
+
+OpenFGA
+-------
+JIMM uses OpenFGA for authorisation and currently, the OpenFGA charm does not support TLS.
```
Comment (SimoneDutto) [inline how-to/security_hardening.rst]: a link to openfga charm  or openfga itself would be good

File: how-to/security_hardening.rst [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+Vault
+-----
+TLS is enabled by default when communicating with the Vault charm.
```
Comment (babakks) [inline how-to/security_hardening.rst]: Maybe a link to the Vault charm?

File: how-to/security_hardening.rst [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
```
Comment (SimoneDutto) [inline how-to/security_hardening.rst]: link to vault charm

File: how-to/security_hardening.rst [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+JIMM uses Vault for storing cloud credentials, JWKS, and other secrets.
+
+Juju Controllers
+----------------
+TLS is enabled by default when communicating with controllers.
+
+When adding a Juju controller to JIMM, the self signed certificate of the controller is given to
```
Comment (babakks) [inline how-to/security_hardening.rst]: s/self signed/self-signed

File: how-to/security_hardening.rst [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+JIMM.
+
+
+PostgreSQL
+----------
+JIMM uses PostgreSQL as its persistent storage layer. The communication between PostgreSQL can be encrypted
```
Comment (babakks) [inline how-to/security_hardening.rst]: s/between/with

File: how-to/security_hardening.rst [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+via TLS. To enable TLS for charmed PostgreSQL you can follow this `guide <https://charmhub.io/postgresql-k8s/docs/t-enable-tls?channel=14/stable>`__.
+
+    As of October 2024, you need to manually restart JIMM if you enable TLS on PostgreSQL after having related the JIMM and PostgreSQL charms.
```
Comment (babakks) [inline how-to/security_hardening.rst]: Please surround this with a tip box.

File: how-to/security_hardening.rst [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+-----
+TLS is enabled by default when communicating with the Vault charm.
+
+JIMM uses Vault for storing cloud credentials, JWKS, and other secrets.
+
+Juju Controllers
+----------------
+TLS is enabled by default when communicating with controllers.
+
+When adding a Juju controller to JIMM, the self signed certificate of the controller is given to
```
Comment (SimoneDutto) [inline how-to/security_hardening.rst]: we should have another doc explain how a new controller is added to jaas, link it here.
If we don't have it ignore this but TLDR explain how a controller "is added"

File: .custom_wordlist.txt (hunk 6 above)
Comment (ale8k) [reply .custom_wordlist.txt]: oh why?

File: how-to/security_hardening.rst (hunk 9 above)
Comment (ale8k) [reply how-to/security_hardening.rst]: ooh

File: how-to/security_hardening.rst [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+these rules can be hardened via TLS and the charm provides a means to do so. See `here <https://charmhub.io/nginx-ingress-integrator>`__.
+
+Our LEGO charms provide certificates for charms from a desired ACME server and can be integrated
+with the integrator to enable TLS at the ingress level. See `here <https://charmhub.io/httprequest-lego-k8s>`__.
+
+You will require a domain that your ACME is aware of and an NGINX ingress controller installed
+on your Kubernetes cluster.
+
+With JAAS deployed, you can deploy both LEGO and the integrator, and integrate your LEGO charm deployment
+to your ingress integrator, and then the ingress integrator to JIMM to enable TLS ingress for
```
Comment (SimoneDutto) [inline how-to/security_hardening.rst]: for?

File: how-to/security_hardening.rst (hunk 18 above)
Comment (babakks) [reply how-to/security_hardening.rst]: Like this:

```suggestion
.. hint::
    As of October 2024, you need to manually restart JIMM if you enable TLS on PostgreSQL after having related the JIMM and PostgreSQL charms.
```

File: how-to/security_hardening.rst [hunk 21]
```diff
@@ -0,0 +1,68 @@
+JAAS: Security Hardening
+========================
+JIMM, the service at the centre of JAAS can be hardened in a number of ways. This
```
Comment (SimoneDutto) [inline how-to/security_hardening.rst:3]: I think you should mention the security overview doc somewhere to give the reader more context on where we stand security-wise

Comment (babakks) [reply how-to/security_hardening.rst:3]: I agree with this. You can do it with a hint box:

```
.. hint::
    As a reference on JAAS security overview, check out :doc:`this <../reference/security>` topic.
```

File: how-to/security_hardening.rst [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+a mechanism by which browsers make a "preflight" request to the server hosting the cross-origin 
+resource, in order to check that the server will permit the actual request. In that preflight, 
+the browser sends headers that indicate the HTTP method and headers that will be used in the 
+actual request.
+
+To set CORS on JIMM, use the configuration option "cors-allowed-origins".
+
+Ingress TLS
+-----------
+The NGINX Ingress Integrator is a a charm responsible for creating Kubernetes ingress rules,
```
Comment (SimoneDutto) [inline how-to/security_hardening.rst]: i feel like this section is very "heavy", we should add another doc page with "how to enable TLS for ingress" like we have for livepatch, and here just the link to  it

File: how-to/security_hardening.rst (hunk 19 above)
Comment (ale8k) [reply how-to/security_hardening.rst]: we have one ill hint it again

File: how-to/security_hardening.rst (hunk 22 above)
Comment (ale8k) [reply how-to/security_hardening.rst]: I kinda like this how it is, as it's clear on what I do to get public certs into my ingress. But i agree that maybe its sort of not the right place as this is kind of a default thing everybody would do and not just a hardening activity?

Comment (SimoneDutto) [reply how-to/security_hardening.rst]: yes, exactly I see this as a good separate HOW-TO.
Happy to approve this PR as it easy but I think we should add a TLS ingress how to

**Review by SimoneDutto — DISMISSED** [review]
good job, writing doc is hard

File: how-to/security_hardening.rst (hunk 22 above)
Comment (ale8k) [reply how-to/security_hardening.rst]: i agree, or maybe this links the how to? ill do that instead and make a "TLS" how to

File: how-to/security_hardening.rst [hunk 23]
```diff
@@ -3,65 +3,63 @@ JAAS: Security Hardening
...
 Juju Controllers
 ----------------
 TLS is enabled by default when communicating with controllers.
 
-When adding a Juju controller to JIMM, the self signed certificate of the controller is given to
+When adding a Juju controller to JIMM, the self-signed certificate of the controller is given to
 JIMM.
 
+.. hint::  
+    Checkout :doc:`this <../how-to/add_controller>` topic for adding controllers to JAAS.
```
Comment (babakks) [inline how-to/security_hardening.rst]: You're already in the `how-to` directory. So, just link to `./add_controller`.

File: how-to/setup_ingress_with_tls.rst [hunk 24]
```diff
@@ -0,0 +1,13 @@
+JAAS: Setup Ingress with TLS
+============================
+The NGINX Ingress Integrator is a a charm responsible for creating Kubernetes ingress rules,
```
Comment (babakks) [inline how-to/setup_ingress_with_tls.rst]: s/a a/a

**Review by babakks — APPROVED** [review]
With just a few changes.

File: how-to/security_hardening.rst (hunk 23 above)
Comment (ale8k) [reply how-to/security_hardening.rst]: oh lol

File: how-to/security_hardening.rst [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+a mechanism by which browsers make a "pre-flight" request to the server hosting the cross-origin 
+resource, in order to check that the server will permit the actual request. In that pre-flight, 
+the browser sends headers that indicate the HTTP method and headers that will be used in the 
+actual request.
+
+To set CORS on JIMM, use the configuration option ``cors-allowed-origins``.
+
+Ingress TLS
+-----------
+Please refer :doc:`here <../how-to/setup_ingress_with_tls>`.
```
Comment (babakks) [inline how-to/security_hardening.rst]: And this one.

**Review by babakks — DISMISSED** [review]

File: how-to/security_hardening.rst (hunk 23 above)
Comment (babakks) [reply how-to/security_hardening.rst]: There's one more (check my next comment).

File: how-to/security_hardening.rst (hunk 25 above)
Comment (ale8k) [reply how-to/security_hardening.rst]: doen!

**Review by babakks — APPROVED** [review]
Great work. :heart:

**Review by SimoneDutto — APPROVED** [review]

**Pull request merged**